### ISSUE TITLE
Better specify what an unknown type credential descriptor being ignored means

### ISSUE THREAD

**Issue by nsatragno** [w3c/webauthn#1748]
We use credential descriptors "`PublicKeyCredentialDescriptor`" for allow lists & exclude lists. The spec says

> [...][client platforms](https://w3c.github.io/webauthn/#client-platform) MUST ignore any [PublicKeyCredentialDescriptor](https://w3c.github.io/webauthn/#dictdef-publickeycredentialdescriptor) with an unknown [type](https://w3c.github.io/webauthn/#dom-publickeycredentialdescriptor-type).

There's no further specification for unknown types. If we follow the spec to the letter, a possible interpretation ends up with:

```javascript
navigator.credentials.get({
  publicKey: {
    allowCredentials: [{type: "not-yet-supported-by-browser", id: <id>}],
  }
});
```

Resulting in an empty `allowCredentials` request, which is definitely not what the RP intended. I think we should immediately return `NotAllowedError` in this case, based on prior text:

> If issuedRequests is empty, options.[allowCredentials](https://w3c.github.io/webauthn/#dom-publickeycredentialrequestoptions-allowcredentials) is not empty, and no authenticator will become available for any [public key credentials](https://w3c.github.io/webauthn/#public-key-credential) therein,
Indicate to the user that no eligible credential could be found. When the user acknowledges the dialog, throw a "[NotAllowedError](https://webidl.spec.whatwg.org/#notallowederror)" [DOMException](https://webidl.spec.whatwg.org/#idl-DOMException).

This is not a problem for `excludeCredentials`, since in that case the browser will simply discard the unknown credential type.

**Comment by Firstyear** [comment]
This a good analysis, I agree with this :)

**Comment by emlun** [comment]
Interesting point. I agree that implicitly falling back to behaving as if `allowCredentials` were empty is clearly not what the RP intended, since that entails username-less authentication, but I also think that similar to #1738, ignoring unknown values is needed for forward compatibility.

Maybe we can change the requirement to be that "[client platforms](https://w3c.github.io/webauthn/#client-platform) MUST ignore any [PublicKeyCredentialDescriptor](https://w3c.github.io/webauthn/#dictdef-publickeycredentialdescriptor) with an unknown [type](https://w3c.github.io/webauthn/#dom-publickeycredentialdescriptor-type), treating the item as if it was not present" ***but*** if this results in an empty `allowCredentials`, then throw an error? In that case we would need to move the requirement from the `PublicKeyCredentialDescriptor` definition to the definitions of `excludeCredentials` and `allowCredentials` (and/or maybe the `create()` and `get()` operations).

**Comment by nsatragno** [comment]
> I also think that similar to https://github.com/w3c/webauthn/issues/1738, ignoring unknown values is needed for forward compatibility.

Yes, this came up when fixing that issue on chrome.

> Maybe we can change the requirement to be that "[client platforms](https://w3c.github.io/webauthn/#client-platform) MUST ignore any [PublicKeyCredentialDescriptor](https://w3c.github.io/webauthn/#dictdef-publickeycredentialdescriptor) with an unknown [type](https://w3c.github.io/webauthn/#dom-publickeycredentialdescriptor-type), treating the item as if it was not present" but if this results in an empty allowCredentials, then throw an error?

Agreed.

**Comment by agl** [comment]
Addressed in #1971
